server/service/mdm.go

```diff
@@ -404,3 +412,234 @@ func (svc *Service) VerifyMDMWindowsConfigured(ctx context.Context) error {
+// Apple or Windows MDM Middleware
+////////////////////////////////////////////////////////////////////////////////
+
+func (svc *Service) VerifyMDMAppleOrWindowsConfigured(ctx context.Context) error {
+	appCfg, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		// skipauth: Authorization is currently for user endpoints only.
+		svc.authz.SkipAuthorization(ctx)
+		return err
+	}
+
+	// Apple or Windows MDM configuration setting
+	if !appCfg.MDM.EnabledAndConfigured && !appCfg.MDM.WindowsEnabledAndConfigured {
+		// skipauth: Authorization is currently for user endpoints only.
+		svc.authz.SkipAuthorization(ctx)
+		return fleet.ErrMDMNotConfigured
+	}
+
+	return nil
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Run Apple or Windows MDM Command
+////////////////////////////////////////////////////////////////////////////////
+
+type runMDMCommandRequest struct {
+	Command   string   `json:"command"`
+	HostUUIDs []string `json:"host_uuids"`
+}
+
+type runMDMCommandResponse struct {
+	*fleet.CommandEnqueueResult
+	Err error `json:"error,omitempty"`
+}
+
+func (r runMDMCommandResponse) error() error { return r.Err }
+
+func runMDMCommandEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*runMDMCommandRequest)
+	result, err := svc.RunMDMCommand(ctx, req.Command, req.HostUUIDs)
+	if err != nil {
+		return runMDMCommandResponse{Err: err}, nil
+	}
+	return runMDMCommandResponse{
+		CommandEnqueueResult: result,
+	}, nil
+}
+
+func (svc *Service) RunMDMCommand(ctx context.Context, rawBase64Cmd string, hostUUIDs []string) (result *fleet.CommandEnqueueResult, err error) {
+	hosts, err := svc.authorizeAllHostsTeams(ctx, hostUUIDs, false, fleet.ActionWrite, &fleet.MDMCommandAuthz{})
+	if err != nil {
+		return nil, err
+	}
+	if len(hosts) == 0 {
+		err := fleet.NewInvalidArgumentError("host_uuids", "No hosts targeted. Make sure you provide a valid UUID.").WithStatus(http.StatusNotFound)
+		return nil, ctxerr.Wrap(ctx, err, "no host received")
+	}
+
+	platforms := make(map[string]bool)
+	for _, h := range hosts {
+		if !h.MDMInfo.IsFleetEnrolled() {
+			err := fleet.NewInvalidArgumentError("host_uuids", "Can't run the MDM command because one or more hosts have MDM turned off. Run the following command to see a list of hosts with MDM on: fleetctl get hosts --mdm.").WithStatus(http.StatusPreconditionFailed)
+			return nil, ctxerr.Wrap(ctx, err, "check host mdm enrollment")
+		}
+		platforms[h.FleetPlatform()] = true
+	}
+	if len(platforms) != 1 {
+		err := fleet.NewInvalidArgumentError("host_uuids", "All hosts must be on the same platform.")
+		return nil, ctxerr.Wrap(ctx, err, "check host platform")
+	}
+
+	// it's a for loop but at this point it's guaranteed that the map has a single value.
+	var commandPlatform string
+	for platform := range platforms {
+		commandPlatform = platform
+	}
```

Comment:
just out of curiosity: why not grab the first one using an index?

Reply:
Ah it's because this is a map, not a slice, so we don't know what index we're looking for. We know at this point it has only a single value in the map, so we loop and grab the first key.